pyro/ops/sumproduct.py

```diff
@@ -16,10 +16,31 @@ def zip_align_right(xs, ys):
 
 
 def memoized_squeeze(tensor):
-    if hasattr(tensor, '_memoized_squeeze'):
-        return tensor._memoized_squeeze
+    """
+    Computes ``tensor.squeeze()`` memoizing the result for the lifetime of
+    ``tensor``. This enables sharing when used inside
+    :func:`~opt_einsum.shared_intermediates`.
+    """
+    if hasattr(tensor, '_pyro_memoized_squeeze'):
+        return tensor._pyro_memoized_squeeze
     result = tensor.squeeze()
-    tensor._memoized_squeeze = result
+    tensor._pyro_memoized_squeeze = result
+    return result
+
+
+def memoized_sum_keepdim(tensor, dim):
```

Comment:
These memoized functions are needed to improve sharing via `shared_intermediates`.